### ISSUE TITLE
Question about the fast tokenizer in pi05

### ISSUE THREAD

**Issue by Tgzz666** [Physical-Intelligence/openpi#708]
After reading the paper, I understand that PI05 uses mixed training on both multi-modal web data and robotics data. However, when encoding fast action tokens, it still uses the <location> token from PaliGemma. I’d like to ask: when training on grounding QA tasks from multi-modal web data, is it really appropriate to still use the <location> token? Doesn’t this cause confusion or conflict?

**Comment by kpertsch** [comment]
We use a range of rarely used tokens for the FAST action tokens, so there is no conflict with the web data targets.

**Comment by Tgzz666** [comment]
> We use a range of rarely used tokens for the FAST action tokens, so there is no conflict with the web data targets.

@kpertsch Sorry for being unclear earlier. My concern is: in Pi-Fast, the fast-action tokens occupy PaliGemma’s <loc> token slots. According to the Pi-0.5 paper, the mixed web VQA data still contains grounding QAs. Don’t these two choices conflict? Or did Pi-0.5 introduce a new token ID to represent fast-action tokens instead of hijacking the <loc> ones?

**Comment by kpertsch** [comment]
Oh this is what I meant: we DO use existing tokens in the token range, but we use rarely used tokens in Paligemma's vocab (typically obscure unicode characters), so in practice there is no interference with tokens that frequently appear in VQA datasets.

**Comment by Tgzz666** [comment]
> Oh this is what I meant: we DO use existing tokens in the token range, but we use rarely used tokens in Paligemma's vocab (typically obscure unicode characters), so in practice there is no interference with tokens that frequently appear in VQA datasets.

@kpertsch I see — you’re saying that the fast-action token ID ranges in Pi-0.5 and Pi-Fast are different, which clears up my confusion entirely. Thank you very much for that!
I have a small request: I’m trying to reproduce the Pi-0.5 training code myself, but the loss on the fast tokens is exploding, and I suspect the token IDs I’m using don’t match those seen during pre-training.
Would it be possible for you to share the exact token-ID range used for fast actions in Pi-0.5? It would be an enormous help for me.

**Comment by kpertsch** [comment]
```
NUM_DCT_TOKENS = 2048
NUM_LOC_TOKENS = 1024
NUM_SEG_TOKENS = 128
NUM_DCT_PAD_TOKENS = 32
GEMMA_TOKENIZER_SIZE = 256_000
PALIGEMMA_TOKENIZER_SIZE = GEMMA_TOKENIZER_SIZE + NUM_LOC_TOKENS + NUM_SEG_TOKENS
PAD_LOC_SEG_TOKEN_RANGE = (
    PALIGEMMA_TOKENIZER_SIZE - (NUM_SEG_TOKENS + NUM_LOC_TOKENS + NUM_DCT_PAD_TOKENS),
    PALIGEMMA_TOKENIZER_SIZE,
)
DCT_TOKEN_RANGE = (PAD_LOC_SEG_TOKEN_RANGE[0] - NUM_DCT_TOKENS, PAD_LOC_SEG_TOKEN_RANGE[0])
```

**Comment by wjydym** [comment]
> > Oh this is what I meant: we DO use existing tokens in the token range, but we use rarely used tokens in Paligemma's vocab (typically obscure unicode characters), so in practice there is no interference with tokens that frequently appear in VQA datasets.
> 
> [@kpertsch](https://github.com/kpertsch) I see — you’re saying that the fast-action token ID ranges in Pi-0.5 and Pi-Fast are different, which clears up my confusion entirely. Thank you very much for that! I have a small request: I’m trying to reproduce the Pi-0.5 training code myself, but the loss on the fast tokens is exploding, and I suspect the token IDs I’m using don’t match those seen during pre-training. Would it be possible for you to share the exact token-ID range used for fast actions in Pi-0.5? It would be an enormous help for me.

Have you finished yet? @Tgzz666

**Comment by Tgzz666** [comment]
> > > Oh this is what I meant: we DO use existing tokens in the token range, but we use rarely used tokens in Paligemma's vocab (typically obscure unicode characters), so in practice there is no interference with tokens that frequently appear in VQA datasets.
> > 
> > 
> > [@kpertsch](https://github.com/kpertsch) I see — you’re saying that the fast-action token ID ranges in Pi-0.5 and Pi-Fast are different, which clears up my confusion entirely. Thank you very much for that! I have a small request: I’m trying to reproduce the Pi-0.5 training code myself, but the loss on the fast tokens is exploding, and I suspect the token IDs I’m using don’t match those seen during pre-training. Would it be possible for you to share the exact token-ID range used for fast actions in Pi-0.5? It would be an enormous help for me.
> 
> Have you finished yet? [@Tgzz666](https://github.com/Tgzz666)

I’m sorry — I didn’t, because I worry that following the description above would make the fast token overlap with the loc token. Wouldn’t that introduce ambiguity?

**Comment by smartparrot** [comment]
how the fast action tokens corresponds to the action head's continuous action chunk? 
Additionally, I guess one subtask predicted need multiple action chunks to complete it, one fast action tokens can represent all the action chunks needed to complete the subtask ?
Thanks

**Comment by HyeonbeomChoi** [comment]
@kpertsch It appears that the token range needs to be revised to align with the implementation in src/openpi/models/tokenizer.py.

In _act_tokens_to_paligemma_tokens, the token mapping is calculated as follows:

```
def _act_tokens_to_paligemma_tokens(self, tokens: np.ndarray | list[int]) -> np.ndarray:
        if isinstance(tokens, list):
            tokens = np.array(tokens)
        return self._paligemma_tokenizer.vocab_size() - 1 - self._fast_skip_tokens - tokens
```

Given that the vocab_size is defined as 256_000 + 1024 + 128 and self._fast_skip_tokens is set to 128 (matching NUM_SEG_TOKENS), the ranges should likely be calculated without NUM_DCT_PAD_TOKENS.

I suggest the following adjustment: 

> ```
> NUM_DCT_TOKENS = 2048
> NUM_LOC_TOKENS = 1024
> NUM_SEG_TOKENS = 128
> # NUM_DCT_PAD_TOKENS = 32
> GEMMA_TOKENIZER_SIZE = 256_000
> PALIGEMMA_TOKENIZER_SIZE = GEMMA_TOKENIZER_SIZE + NUM_LOC_TOKENS + NUM_SEG_TOKENS
> SEG_TOKEN_RANGE = (
>     PALIGEMMA_TOKENIZER_SIZE - (NUM_SEG_TOKENS),
>     PALIGEMMA_TOKENIZER_SIZE,
> )
> DCT_TOKEN_RANGE = (SEG_TOKEN_RANGE[0] - NUM_DCT_TOKENS, SEG_TOKEN_RANGE[0]) # 2048 tokens for pi0_fast using FASTTokenizer
> ```

Thanks.

**Comment by Tgzz666** [comment]
> ```
> NUM_DCT_TOKENS = 2048
> NUM_LOC_TOKENS = 1024
> NUM_SEG_TOKENS = 128
> NUM_DCT_PAD_TOKENS = 32
> GEMMA_TOKENIZER_SIZE = 256_000
> PALIGEMMA_TOKENIZER_SIZE = GEMMA_TOKENIZER_SIZE + NUM_LOC_TOKENS + NUM_SEG_TOKENS
> PAD_LOC_SEG_TOKEN_RANGE = (
>     PALIGEMMA_TOKENIZER_SIZE - (NUM_SEG_TOKENS + NUM_LOC_TOKENS + NUM_DCT_PAD_TOKENS),
>     PALIGEMMA_TOKENIZER_SIZE,
> )
> DCT_TOKEN_RANGE = (PAD_LOC_SEG_TOKEN_RANGE[0] - NUM_DCT_TOKENS, PAD_LOC_SEG_TOKEN_RANGE[0])
> ```

@kpertsch 
Hello, after several months, I still have some questions about this that are unclear to me, and I hope to consult with you. You previously mentioned that NUM_DCT_PAD_TOKENS is used for padding the length of action tokens. I'm quite curious—shouldn't we just directly use PaliGemma's pad token (token_id: 0) for padding? If we're using a new pad token, shouldn't just one be sufficient? Why are there 32 of them? Thank you for your generosity, I truly appreciate it.

**Comment by varunkumar3618** [comment]
I think that NUM_DCT_PAD_TOKENS is used because the start of image token is in the range, not for padding to a longer size.
